[CT-402] Randomly failing test: test_hard_delete_snapshot.py

The failing test seems to be a 1 second difference between 2 timestamps in tests/functional/simple_snapshot/test_hard_delete_snapshot.py just for one Python/ OS combo

            )
E           assert datetime.datetime(2022, 3, 21, 19, 39, 23, tzinfo=<UTC>) >= datetime.datetime(2022, 3, 21, 19, 39, 24, tzinfo=<UTC>)
E            +  where datetime.datetime(2022, 3, 21, 19, 39, 23, tzinfo=<UTC>) = <built-in method replace of datetime.datetime object at 0x7f0793efc780>(tzinfo=<UTC>)
E            +    where <built-in method replace of datetime.datetime object at 0x7f0793efc780> = datetime.datetime(2022, 3, 21, 19, 39, 23).replace
E            +    and   <UTC> = pytz.UTC
E            +  and   datetime.datetime(2022, 3, 21, 19, 39, 24, tzinfo=<UTC>) = <built-in method replace of datetime.datetime object at 0x7f0794198900>(microsecond=0)
E            +    where <built-in method replace of datetime.datetime object at 0x7f0794198900> = datetime.datetime(2022, 3, 21, 19, 39, 24, 2693, tzinfo=<UTC>).replace

/home/runner/work/dbt-core/dbt-core/tests/functional/simple_snapshot/test_hard_delete_snapshot.py:146: AssertionError

--- a/.changes/unreleased/Fixes-20220408-155512.yaml
+++ b/.changes/unreleased/Fixes-20220408-155512.yaml
@@ -1,0 +1,7 @@
+kind: Fixes
+body: Fix hard delete snapshot test
+time: 2022-04-08T15:55:12.552138-04:00
+custom:
+  Author: gshank
+  Issue: "4916"
+  PR: "5020"

--- a/tests/functional/simple_snapshot/test_hard_delete_snapshot.py
+++ b/tests/functional/simple_snapshot/test_hard_delete_snapshot.py
@@ -46,6 +46,7 @@ def test_snapshot_hard_delete(project):
     path = os.path.join(project.test_data_dir, "seed_pg.sql")
     project.run_sql_file(path)
 
+    # run the first snapshot
     datetime_snapshot()
 
     check_relations_equal(project.adapter, ["snapshot_expected", "snapshot_actual"])
@@ -81,7 +82,8 @@ def test_snapshot_hard_delete(project):
         assert result[-1].replace(tzinfo=pytz.UTC) >= invalidated_snapshot_datetime
 
     # revive records
-    revival_timestamp = datetime.now(pytz.UTC).strftime(r"%Y-%m-%d %H:%M:%S")
+    # Timestamp must have microseconds for tests below to be meaningful
+    revival_timestamp = datetime.now(pytz.UTC).strftime("%Y-%m-%d %H:%M:%S.%f")
     project.run_sql(
         """
         insert into {}.{}.seed (id, first_name, last_name, email, gender, ip_address, updated_at) values
@@ -93,9 +95,11 @@ def test_snapshot_hard_delete(project):
     )
 
     # snapshot and assert records were revived
+    # Note: the revived_snapshot_datetime here is later than the revival_timestamp above
     revived_snapshot_datetime = datetime_snapshot()
 
     # records which weren't revived (id != 10, 11)
+    # dbt_valid_to is not null
     invalidated_records = project.run_sql(
         """
         select
@@ -117,6 +121,7 @@ def test_snapshot_hard_delete(project):
         assert result[1].replace(tzinfo=pytz.UTC) >= invalidated_snapshot_datetime
 
     # records which were revived (id = 10, 11)
+    # dbt_valid_to is null
     revived_records = project.run_sql(
         """
         select
@@ -135,10 +140,8 @@ def test_snapshot_hard_delete(project):
     assert len(revived_records) == 2
     for result in revived_records:
         # result is a tuple, the dbt_valid_from is second and dbt_valid_to is latest
+        # dbt_valid_from is the same as the 'updated_at' added in the revived_rows
+        # dbt_valid_to is null
         assert isinstance(result[1], datetime)
-        # there are milliseconds (part of microseconds in datetime objects) in the
-        # revived_snapshot_datetime and not in result datetime so set the microseconds to 0
-        assert result[1].replace(tzinfo=pytz.UTC) >= revived_snapshot_datetime.replace(
-            microsecond=0
-        )
+        assert result[1].replace(tzinfo=pytz.UTC) <= revived_snapshot_datetime
         assert result[2] is None

